server/src/main/java/org/apache/druid/client/MetadataSegmentView.java

```diff
@@ -0,0 +1,202 @@
+        segmentWatcherConfig.getWatchedDataSources()
+    );
+
+    final DateTime timestamp = DateTimes.nowUtc();
+    while (metadataSegments.hasNext()) {
+      final DataSegment interned = DataSegmentInterner.intern(metadataSegments.next());
+      // timestamp is used to filter deleted segments
+      publishedSegments.put(interned, timestamp);
+    }
+    // filter the segments from cache whose timestamp is not equal to latest timestamp stored,
+    // since the presence of a segment with an earlier timestamp indicates that
+    // "that" segment is not returned by coordinator in latest poll, so it's
+    // likely deleted and therefore we remove it from publishedSegments
+    publishedSegments.entrySet().removeIf(e -> e.getValue() != timestamp);
+
+  }
+
+  public Iterator<DataSegment> getPublishedSegments()
+  {
+    return publishedSegments.keySet().iterator();
+  }
+
+  // Note that coordinator must be up to get segments
+  private static JsonParserIterator<DataSegment> getMetadataSegments(
+      DruidLeaderClient coordinatorClient,
+      ObjectMapper jsonMapper,
+      BytesAccumulatingResponseHandler responseHandler,
+      Set<String> watchedDataSources
+  )
+  {
+    String query = "/druid/coordinator/v1/metadata/segments";
+    if (watchedDataSources != null && !watchedDataSources.isEmpty()) {
+      log.debug(
+          "filtering datasources in published segments based on broker's watchedDataSources[%s]", watchedDataSources);
+      final StringBuilder sb = new StringBuilder();
+      for (String ds : watchedDataSources) {
+        sb.append("datasources=" + ds + "&");
+      }
+      sb.setLength(sb.length() - 1);
+      query = "/druid/coordinator/v1/metadata/segments?" + sb;
+    }
+    Request request;
+    try {
+      request = coordinatorClient.makeRequest(
+          HttpMethod.GET,
+          StringUtils.format(query),
+          false
+      );
+    }
+    catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+    ListenableFuture<InputStream> future = coordinatorClient.goAsync(
+        request,
+        responseHandler
+    );
+
+    final JavaType typeRef = jsonMapper.getTypeFactory().constructType(new TypeReference<DataSegment>()
+    {
+    });
+    return new JsonParserIterator<>(
+        typeRef,
+        future,
+        request.getUrl().toString(),
+        null,
+        request.getUrl().getHost(),
+        jsonMapper,
+        responseHandler
+    );
+  }
+
+  private class PollTask implements Runnable
+  {
+    @Override
+    public void run()
+    {
+      long delayMS = POLL_PERIOD_IN_MS;
+      try {
+        long pollStartTime = System.nanoTime();
+        poll();
+        long pollEndTime = System.nanoTime();
+        final long pollTimeNs = pollEndTime - pollStartTime;
+        final long pollTimeMs = TimeUnit.NANOSECONDS.toMillis(pollTimeNs);
+        if (pollTimeMs > POLL_PERIOD_IN_MS) {
+          delayMS = 0;
+        } else {
+          delayMS = pollTimeMs;
```

Comment:
`pollTimeMs` is how long `poll` took, so probably `delayMs = Math.max(POLL_PERIOD_IN_MS - pollTimeMs, 0)`

Reply:
changed